### ISSUE TITLE
Request to support W'uau pool heat pumps (Tuya based Wifi device)

### ISSUE THREAD

**Issue by Soepgroenten** [make-all/tuya-local#152]
This seems to be a new brand which is becoming more and more common.

Not much to tell about it, it's working with the rebranded tuya app of w'eau, and the Tuya app itself offcourse.

Below the manual with some parameters specified.

[weau-horizontal-full-inverter-single-phase.pdf](https://github.com/make-all/tuya-local/files/8419410/weau-horizontal-full-inverter-single-phase.pdf)

I'm looking forward for this integration :-)

**Comment by make-all** [comment]
Sorry, but the manual is not sufficient to add support.  Protocol level information is required, and this can only from from the Home Assistant log messages when you try to add it (unsuccessfully presumably), or by monitoring it using another tuya protocol tool like tuya-cli.

**Comment by Soepgroenten** [comment]
> Sorry, but the manual is not sufficient to add support. Protocol level information is required, and this can only from from the Home Assistant log messages when you try to add it (unsuccessfully presumably), or by monitoring it using another tuya protocol tool like tuya-cli.

Thanks for looking into this, I've tried to gather the required information based on your message. I've had no success using Tuya-CLI, whilst I'm very familiar with scripts and modules like the tuya-cli I ended up with some rare issues (but that's off topic)

Out of the Home assistant logging I couldn't get any usefull logging info regarding logs of Tuya-Local, but I've found something that looks pretty usefull at the localtuya logs, please have a look at the attached "Weau_heat_pump_log_DueToLocalTuyaAddIn.txt" file:

[Weau_heat_pump_log_DueToLocalTuyaAddIn.txt](https://github.com/make-all/tuya-local/files/8474857/Weau_heat_pump_log_DueToLocalTuyaAddIn.txt)

I also digged somewhat in the Tuya API and also saw a vew promising results, please have a look a the attached files below.

[Weau_heat_pump_DeviceSpecification_DueToTuyaApi.txt](https://github.com/make-all/tuya-local/files/8474854/Weau_heat_pump_DeviceSpecification_DueToTuyaApi.txt)
[Weau_heat_pump_InstructionSetDueToTuyaApi.txt](https://github.com/make-all/tuya-local/files/8474855/Weau_heat_pump_InstructionSetDueToTuyaApi.txt)
[Weau_heat_pump_InstructionSetOfCategory_DueToTuyaApi.txt](https://github.com/make-all/tuya-local/files/8474856/Weau_heat_pump_InstructionSetOfCategory_DueToTuyaApi.txt)

Where the results in file: "Weau_heat_pump_DeviceSpecification_DueToTuyaApi.txt" looks very familiar with the looks in the Tuya app. See screenshot below for an impression of how the Tuya gui looks like (Android).
![Screenshot_2022-04-12-18-45-07-198](https://user-images.githubusercontent.com/85663848/163013864-1f1c020f-63ed-4492-b3c9-588219a58fe2.jpg)

![Screenshot_2022-04-12-18-45-22-845](https://user-images.githubusercontent.com/85663848/163013859-0e704e8b-a8ea-4848-be53-f2ec966c8090.jpg)

I hope this is something useful you can do your magic with

**Comment by make-all** [comment]
Thanks, those logs and cloud API docs are enough to figure out the fairly simple functionality this heatpump exposes.

**Comment by Soepgroenten** [comment]
> Thanks, those logs and cloud API docs are enough to figure out the fairly simple functionality this heatpump exposes.

Good to hear the provided information is sufficient, and that the heat pump seems to be pretty easy to implement.

Great thanks in advance 👍

**Comment by make-all** [comment]
There is some inconsistency between the TuyaAPI docs and the logs, screenshots and manual.

In the two API docs for the device (the category doc obviously contains things not relevant for this device, so I am discarding that), they both say the valid modes are "cold" and "hot", mapping to "Eco Cool" and "Boost Heat" respectively.

But in the log, the mode is "auto", and in the screenshot, there are modes "Boost Heat", "Eco Cool", "Auto", and "Eco Heat" visible.  And in the manual, it says the valid modes are "Boost Heat", "Eco Heat", "Cool" and "Auto".

So I think to get the values correct, I need you to collect the logs for each of the modes.

There is also some inconsistency in that the manual says that the App functions include Timer and also Error display.  But there is only one left over DP (6), so only one of those functions is possible.  For now, I'll leave it as an unknown attribute so it can be observed.  But if you can figure out what it is, it may be able to provide additional useful functionality (particularly if it is error codes, as it can be useful to trigger automations such as sending alerts off those).

**Comment by fwelvering** [comment]
So i have this W'eau heatpump but when adding it to HA is try's to add an 'Simple switch'
I added the W'eau filt to the devices folder in HA and restarted HA to be save.

Is there anything i should do to get it working?

**Comment by make-all** [comment]
The support was not yet in the released version, until the version released today.

**Comment by make-all** [comment]
It seems there are different models of this brand using completely different protocols.  Do you have the model number?
And if you have access to iot.tuya.com (to get your local key perhaps, or to set up the cloud integration), then under Cloud / API Explorer / Device Control can you run the last Get Device Specification Attribute function with your data center and device id.  This will help with the function and range of some of those values.

**Comment by fwelvering** [comment]
Hi,

The output from the Get Device Specification:
`{
	"result": {
		"category": "rs",
		"functions": [{
			"code": "switch",
			"type": "Boolean",
			"values": "{}"
		}, {
			"code": "mode",
			"type": "Enum",
			"values": "{\"range\":[\"auto\"]}"
		}, {
			"code": "temp_set",
			"type": "Integer",
			"values": "{\"unit\":\"℃\",\"min\":-500,\"max\":1000,\"scale\":0,\"step\":1}"
		}],
		"status": [{
			"code": "switch",
			"type": "Boolean",
			"values": "{}"
		}, {
			"code": "mode",
			"type": "Enum",
			"values": "{\"range\":[\"auto\"]}"
		}, {
			"code": "temp_set",
			"type": "Integer",
			"values": "{\"unit\":\"℃\",\"min\":-500,\"max\":1000,\"scale\":0,\"step\":1}"
		}, {
			"code": "temp_current",
			"type": "Integer",
			"values": "{\"unit\":\"℃\",\"min\":-1000,\"max\":1000,\"scale\":1,\"step\":1}"
		}, {
			"code": "fault",
			"type": "bitmap",
			"values": "{\"label\":[\"sys_high_fault\",\"sys_low_fault\",\"flow_fault\",\"power_fault\",\"cooling_fault\",\"heating_fault\",\"temp_dif_fault\",\"in_temp_fault\",\"eff_temp_fault\",\"coil_temp_fault\",\"ret_temp_fault\",\"news_fault\",\"amb_temp_fault\",\"lack_water\",\"serious_fault\",\"sensor_fault\",\"motor_fault\"],\"maxlen\":16}"
		}]
	},
	"success": true,
	"t": 1654852071365,
	"tid": "ce28efa6e89c11ec87342aa367714e21"
}
`
And from device details (for model and product no)
`
{
	"result": {
		"active_time": 1653652854,
		"biz_type": 0,
		"category": "rs",
		"create_time": 1652620325,
		"icon": "smart/icon/ay1513418289516FAXSV/df2cf842ab71c359308c3814492d5ee9.png",
		"id": "70631187e868e7d4ebde",
		"ip": "<redacted>",
		"lat": "<redacted>",
		"local_key": "<redacted>",
		"lon": "<redacted>",
		"model": "RCDF211344",
		"name": "Weau pool heat pump",
		"online": false,
		"owner_id": "<redacted>",
		"product_id": "zqwet1zcvj2wcxzw",
		"product_name": "Weau pool heat pump",
		"status": [{
			"code": "switch",
			"value": false
		}, {
			"code": "mode",
			"value": "eheat"
		}, {
			"code": "temp_set",
			"value": 15
		}, {
			"code": "temp_current",
			"value": 259
		}, {
			"code": "fault",
			"value": 0
		}],
		"sub": false,
		"time_zone": "+01:00",
		"uid": "<redacted>",
		"update_time": 1653652859,
		"uuid": "<redacted>"
	},
	"success": true,
	"t": 1654852209433,
	"tid": "20759836e89d11ecaeb54a08c3e1e6f8"
}
`

**Comment by make-all** [comment]
Unfortunately the API Explorer has a lot of similarly named functions which return subtly different results.  The one I was after was the last one listed under the second last category (Device Control).  That is the only one that ties in to the local dp_ids.

**Comment by Soepgroenten** [comment]
I've tested with the current update..to me the integration is a bit of a succes. Unfortunately the devices becomes unavailable over time:
![image](https://user-images.githubusercontent.com/85663848/175032100-15f5bf33-ce6e-4ef9-a4d1-c44b6022ccc1.png)

I Noticed that it only happens when I'm actually looking at the Entity in my dashboard

Reloading Tuya Local doesnt solve the issue. To solve this I have to restart HomeAssistant.

What could this be? And how can I help solving this?

**Comment by make-all** [comment]
Thanks for confirming.  I think the device becoming unavailable is a common issue with Tuya devices - make sure nobody uses the Tuya app, as that can block other local connections to your device. Otherwise, I do see messages like this in the log regularly myself, but they don"t really affect the UI, so I think it is just an occasional poll being missed.

**Comment by fwelvering** [comment]
After installing the latest release mine is still recognised as an simple switch.
Do you need any extra information to hopefully fix the issue?

**Comment by make-all** [comment]
Yes, this issue is still open with the Device Variant tag to reflect that the information you gave in an earlier comment is a variation of the original that will require a new config to handle.

**Comment by make-all** [comment]
I think the device becoming unavailable may be due to the extra undocumented modes it supports.  The documentation you provided specified the valid modes as "hot" and "cold", but the current data you provided had a value of "auto", so I added that also.  Now (and from the screenshot) you are saying there are also Eco heat and Eco cool modes.  fwelvering's current data is showing "eheat", so maybe the eco modes are "eheat" and "ecool"?  Can you confirm?  I think you will get errors from Home Assistant about invalid HVAC modes, since I just pass through unmapped values, but hopefully the error message will have the actual value in it.  Otherwise you may need to enable debug logging for the integration temporarily to  find what values are being set.

**Comment by make-all** [comment]
I need confirmation of the modes supported to complete this support.

**Comment by Soepgroenten** [comment]
I'm sorry for my late reply.
Below the status of each Mode:
**Eco Heat**
![image](https://user-images.githubusercontent.com/85663848/177921942-d6746672-a5be-4f58-830f-70c9b6005ab7.png)
**Boost Heat**
![image](https://user-images.githubusercontent.com/85663848/177921994-f9efbf95-e0bb-4a4b-8913-47b1306a0041.png)
**Eco Cool**
![image](https://user-images.githubusercontent.com/85663848/177922144-f0be6b14-65c6-49e0-9b05-c84a3a673c7e.png)
**Auto**
![image](https://user-images.githubusercontent.com/85663848/177922246-a47ae926-b64c-4044-8c18-4503614fe135.png)

There's also a "waterflow protection" attribute but that one seems to be unfindable in iot.tuya I've tought it was the switch parameter in the responses above but that one doesnt change when the waterflow protection is active.

Below also the latest output of the device specification attributes (with all dp_id's):
{
  "result": {
    "category": "ktkzq",
    "functions": [
      {
        "code": "switch",
        "dp_id": 1,
        "type": "Boolean",
        "values": "{}"
      },
      {
        "code": "temp_set",
        "dp_id": 2,
        "type": "Integer",
        "values": "{\"unit\":\"℃\",\"min\":7,\"max\":60,\"scale\":0,\"step\":1}"
      },
      {
        "code": "mode",
        "dp_id": 4,
        "type": "Enum",
        "values": "{\"range\":[\"hot\",\"cold\"]}"
      }
    ],
    "status": [
      {
        "code": "switch",
        "dp_id": 1,
        "type": "Boolean",
        "values": "{}"
      },
      {
        "code": "temp_set",
        "dp_id": 2,
        "type": "Integer",
        "values": "{\"unit\":\"℃\",\"min\":7,\"max\":60,\"scale\":0,\"step\":1}"
      },
      {
        "code": "temp_current",
        "dp_id": 3,
        "type": "Integer",
        "values": "{\"unit\":\"℃\",\"min\":-300,\"max\":1500,\"scale\":1,\"step\":1}"
      },
      {
        "code": "mode",
        "dp_id": 4,
        "type": "Enum",
        "values": "{\"range\":[\"hot\",\"cold\"]}"
      }
    ]
  },
  "success": true,
  "t": 1657259048137,
  "tid": "fc22c9d8fe8011ec8201daf929f70cab"
}

In the end I must say, the device works well now for a while already...Only the eco heat and waterflow protection is missing.

The device also didnt stopped reporting anymore for some reason...so it's reliable now and I can depend on it.

**Comment by make-all** [comment]
Waterflow protection is usually a bit on a more general fault code dps.  It will be one of the as yet unidentified ones I think.

**Comment by Soepgroenten** [comment]
> integer fault code

I See, fwelvering got these info from the Get Device Specification query, but mine doesn't state anything like flow_fault or anything that looks like it. probably they do it by magic with my heat pump 😁

Just like the Eco mode.

If the Eco heat mode and waterflow switch aren't possible with Tuya Local then so be it 😉 I'm already glad it works like the way it does now and it's running smoothly for weeks now so we can Mark this one as completed.

**Comment by make-all** [comment]
eco should work, I think, it is just the documentation that is lacking.  I suspect from the initial information provided that dps 6 will be the fault and maybe the same values as fwelvering's variant (for which I still need the dp ids to make a config).

**Comment by Soepgroenten** [comment]
Unfortunately I already posted all available DP_id's...there's no othere page than the "Get Device Specification Attribute" under "Device Control".

it states the following:

Get Device Specification Attribute

{
  "result": {
    "category": "ktkzq",
    "functions": [
      {
        "code": "switch",
        "dp_id": 1,
        "type": "Boolean",
        "values": "{}"
      },
      {
        "code": "temp_set",
        "dp_id": 2,
        "type": "Integer",
        "values": "{\"unit\":\"℃\",\"min\":7,\"max\":60,\"scale\":0,\"step\":1}"
      },
      {
        "code": "mode",
        "dp_id": 4,
        "type": "Enum",
        "values": "{\"range\":[\"hot\",\"cold\"]}"
      }
    ],
    "status": [
      {
        "code": "switch",
        "dp_id": 1,
        "type": "Boolean",
        "values": "{}"
      },
      {
        "code": "temp_set",
        "dp_id": 2,
        "type": "Integer",
        "values": "{\"unit\":\"℃\",\"min\":7,\"max\":60,\"scale\":0,\"step\":1}"
      },
      {
        "code": "temp_current",
        "dp_id": 3,
        "type": "Integer",
        "values": "{\"unit\":\"℃\",\"min\":-300,\"max\":1500,\"scale\":1,\"step\":1}"
      },
      {
        "code": "mode",
        "dp_id": 4,
        "type": "Enum",
        "values": "{\"range\":[\"hot\",\"cold\"]}"
      }
    ]
  },
  "success": true,
  "t": 1657267812014,
  "tid": "63cf5a25fe9511ec887d9287a6a8350d"
}

But I do not think this will help you/us further, any clues where to look for more DP_id's? Tried practically every query available in iot.tuya. Also tried debugging but that doesnt show  the dp id's either.

**Comment by make-all** [comment]
The original capture of local data had the following:
```
({'1': False, '2': 33, '3': 195, '4': 'auto', '6': 0}
```
So '6' is the only candidate, I think it is a safe assumption that that is how the water flow fault is being communicated back to the app.  I will make a general "fault" sensor that signals any non-zero value, and also leave the original fault code as an attribute on the climate entity so the information is available in case anyone wants to confirm the meanings of specific values.

**Comment by Soepgroenten** [comment]
> The original capture of local data had the following:
> 
> ```
> ({'1': False, '2': 33, '3': 195, '4': 'auto', '6': 0}
> ```
> 
> So '6' is the only candidate, I think it is a safe assumption that that is how the water flow fault is being communicated back to the app. I will make a general "fault" sensor that signals any non-zero value, and also leave the original fault code as an attribute on the climate entity so the information is available in case anyone wants to confirm the meanings of specific values.

Sounds like a good plan, let me know when the update is available...then I'll test it.

**Comment by make-all** [comment]
Eco mode and fault binary sensor have been added to the original config in 0.18.0.

The second variant is still pending more information on the available modes, as the device log does not seem to match the documentation, and it also does not match any of the modes available in the original version.

**Comment by fwelvering** [comment]
Is there anything you need from me to add the device variant?

**Comment by make-all** [comment]
> Unfortunately the API Explorer has a lot of similarly named functions which return subtly different results. The one I was after was the last one listed under the second last category (Device Control). That is the only one that ties in to the local dp_ids.

Also, the valid modes are listed in the previous API Explorer result as ["auto"], while the current value is "eheat".  I'm afraid you're going to have to do some experimentation to find what the full list of actual actual valid modes is.

**Comment by fwelvering** [comment]
Is this the output you where looking for?

`{
  "result": {
    "category": "rs",
    "functions": [
      {
        "code": "switch",
        "dp_id": 1,
        "type": "Boolean",
        "values": "{}"
      },
      {
        "code": "mode",
        "dp_id": 2,
        "type": "Enum",
        "values": "{\"range\":[\"auto\"]}"
      },
      {
        "code": "temp_set",
        "dp_id": 9,
        "type": "Integer",
        "values": "{\"unit\":\"℃\",\"min\":-500,\"max\":1000,\"scale\":0,\"step\":1}"
      }
    ],
    "status": [
      {
        "code": "switch",
        "dp_id": 1,
        "type": "Boolean",
        "values": "{}"
      },
      {
        "code": "mode",
        "dp_id": 2,
        "type": "Enum",
        "values": "{\"range\":[\"auto\"]}"
      },
      {
        "code": "temp_set",
        "dp_id": 9,
        "type": "Integer",
        "values": "{\"unit\":\"℃\",\"min\":-500,\"max\":1000,\"scale\":0,\"step\":1}"
      },
      {
        "code": "temp_current",
        "dp_id": 10,
        "type": "Integer",
        "values": "{\"unit\":\"℃\",\"min\":-1000,\"max\":1000,\"scale\":1,\"step\":1}"
      },
      {
        "code": "fault",
        "dp_id": 20,
        "type": "bitmap",
        "values": "{\"label\":[\"sys_high_fault\",\"sys_low_fault\",\"flow_fault\",\"power_fault\",\"cooling_fault\",\"heating_fault\",\"temp_dif_fault\",\"in_temp_fault\",\"eff_temp_fault\",\"coil_temp_fault\",\"ret_temp_fault\",\"news_fault\",\"amb_temp_fault\",\"lack_water\",\"serious_fault\",\"sensor_fault\",\"motor_fault\"],\"maxlen\":16}"
      }
    ]
  },
  "success": true,
  "t": 1658318785537,
  "tid": "60d9baac082411eda4ef322e56e3d437"
}`

As for the modes i know ehead and ecool are working, auto is not accepted by the api strangely

**Comment by make-all** [comment]
Second variant is also released in 0.18.1

**Comment by Soepgroenten** [comment]
> Second variant is also released in 0.18.1

Installed, and it looks promising but unfortunately I can't physically check the device as I'm enjoying a wonderfull vacation ☺️.

I will have a look as soon as I'm back at home and let you know

Thnx

**Comment by pergolafabio** [comment]
Hi, also purchased a Weau Mini inverter, what am i doing wrong here? The keys ae correct, since they work with localtuya, but only switch is there discovered, i want the full climate options offcourse!! :-)

also tried other protocols , tried auto also

![image](https://github.com/make-all/tuya-local/assets/43251136/13d95fad-a6b7-4c04-81a7-033aaf7cb5d1)

**Comment by pergolafabio** [comment]
Guys, the current temp is in tuya app , 10 x higher, cause of a wrong decimal, how is this solved here? by a manual dvivide / 10  hardcoded for Weau ?

**Comment by Soepgroenten** [comment]
Personally I think your mini converter depends on another Tuya configuration, so I think your mini converter needs a whole new integration in Tuya local.

2 days ago I successfully re-added my heat pump without a hassle and the most important settings and values are available including the right temperatures, I never've needed additional modifications to show the right temps.



________________________________

**Comment by pergolafabio** [comment]
yeah, i managed now indeed with localtuya, the current temp was 10 x higher, also in Tuya app

But below config is working for me, changed precision to 0.1 , to divide, and used target precision to 1 , otherwise the set temp was also divided by 10

Only thing missing is how to setup the cool mode, allthough i dont use that anyway :-)

```

  "data": {
    "device_config": {
      "friendly_name": "Weau",
      "host": "192.168.0.76",
      "local_key": "XXXXX",
      "protocol_version": "3.4",
      "enable_debug": false,
      "entities": [
        {
          "friendly_name": "Weau",
          "target_temperature_dp": 9,
          "current_temperature_dp": 10,
          "temperature_step": 1,
          "precision": 0.1,
          "hvac_mode_dp": 2,
          "hvac_mode_set": "manual/auto",
          "temperature_unit": "celsius",
          "target_precision": 1,
          "id": 1,
          "platform": "climate"
        }
      ],
```

**Comment by make-all** [comment]
The error is a connection error - possible reasons for those are gathered in discussion #522
Only after the connection error is resolved will you really know whether this is an incompatible model with the existing W'eau support (the exact model I don't know)

**Comment by pergolafabio** [comment]
i always get the connection error , tried restarting the weau, deleted in from the other localtuya, restarted HA, restarted even my  router to drop all connections, nothing helped

now switch back to localtuya instead, and no connection issue, so not sure whats going on with your integration

**Comment by Soepgroenten** [comment]
@make-all 
I'm a happy Tuya Local customer :-) using this integration for over a year with the Tuya based W'eau Pool Heatpump.
This year I'm making more and more use of Voice assistants (Year of the Voice), but unfortunately the Tuya heatpump integration doesn't provide separate entities for the temp set and the current temp. Therefore I'm unable to ask my voice assistant any details of for except: The current temperature of my Pool.

If possible it would be of great benefits to have all attributes listed as separate entities or at least the current temperature would be nice.

Another question:
The heatpump itself shows input temp and output temp on it's display (so the sensors are in place). but Tuya Local and the Tuya app itself doesnt show the input/output temp. Is there any way to have this added to this integration? As it says a lot about the performance of my heatpump and being able to optimise it's efficiency using the gaph's that can be made.

This is prolly not the place for requests like this, but there's no category for Currently Supported device and requests for improvements.

**Comment by pergolafabio** [comment]
For the sensors, why not create them self? you can define template sensors, and then use the correct attribute of your climate?
For the IN/OUT, have a look on the IOT Tuya website, when you debug the DP's for the Weau, do you see them listed there? If not, i dont think its possible, thats propably why you dont see them in the Tuya app either

DO you see them when using the Weau app ?

**Comment by make-all** [comment]
Yes, template sensors is the way to go if you need to split out things that are already covered under other entities into separate sensors.
But voice assistants generally work with climate entities, so you should be able to query the current temperature and set a target temperature without doing this.

**Comment by pergolafabio** [comment]
ah ok,  maybe this one?

https://www.home-assistant.io/integrations/generic_thermostat/

**Comment by make-all** [comment]
This integration already provides the climate entity, generic thermostat is the opposite of what the OP wants; combining discrete switches and sensors into a climate entity.

**Comment by Soepgroenten** [comment]
Thanks for your very quick response guys..I was already looking at creating helpers but template sensors actually seems to do the trick and lowers the complexity a lot :-) I'll have a look into it.

**Comment by pergolafabio** [comment]
if you want, you can post the config, can be usefull :-)

**Comment by Soepgroenten** [comment]
I am using the Integration with a W'eau WFI-007
AL DP's that were available are already available in this integration.

The reason why wanted separate entities is because I can give my entities aliases so in the end it doesnt matter what is beind shout to the voice assistant as I got several members and having to learn what sentences will be recognized  feels a bit unnatural. with aliases I'm able to provide all different possible kinds of recognizable words so this is my mail reason why I want a separate entity...

I've looked into creating a template sensor but I'm unsure which one to use, the one posted here is to set temp's I just want to be able to ask the assistant about the current temperature is being measured in the Pool.

unfortunately I've got only a  little bit of time so finding things out will take some time

@pergolafabio 
What kind of config do you need?
